### DISCUSSION TITLE
support for Binance csv file ?

### DISCUSSION THREAD

**Discussion by jancovic** [provinzio/CoinTaxman#69] [Q&A] [answered]
Hi

Do you support also original Binance transactions csv files without need to use API ?

**Comment by scientes** [comment, marked as answer]
Yes and no, you can just use the binance csv files without using an api key, but we still query price data from binance.

However some operations on binance are not logged in the transactions csv, such as liquid swap reward distribution which can probably only be aquired through the api , so if you have done something which deviates from normal buys/sells there could be some errors

**Comment by jancovic** [comment]
thanks for reply, I am just starting study how it works, to generate gains and losses, e.g. some other tax calculators also support csv transactions file, but for me is question how they calculate the gains only from this file ?  when I look  inside there are only numbers how much sold/bought , but there is no info about current prize, so how they calculate ? for me if I want do my self calculator is better using   spot trade history csv file, where you have all data...
Other question is what are requirements to report taxes, how the report should looks like ?   e.g. if I have running bots ( thousands of transactions), can I also do my own summary without other software  ?  ...just excel sheets with all summary, yes it will be very long sheet :)  


Here is example about buying some ADA from Binanace transactions
UTC_Time	        Account	  Operation	Coin	       Change
6/30/2021 4:48	Spot	          Buy	        ADA	        16.97
6/30/2021 4:48	Spot  	  Buy	        USDT	-22.509008

**Reply by provinzio** [reply]
As scientes said. Additional to the transaction CSV file, we query the corresponding price data from the open binance API (no API key required).

Regarding your last question. It doesn't matter how you calculate your losses/gains, as long as you do it according to tax law.

**Comment by jancovic** [comment]
OK, I tried to run it and I got this error, how can I import my csv files ?

~/CoinTaxman$ python3 src/main.py 
Traceback (most recent call last):
  File "src/main.py", line 20, in <module>
    from book import Book
  File "CoinTaxman/src/book.py", line 26, in <module>
    import misc
  File "CoinTaxman/src/misc.py", line 40, in <module>
    L = TypeVar("L", bound=list[Any])
TypeError: 'type' object is not subscriptable

**Reply by provinzio** [reply]
Check the requirements and your python version. Looks like you use an older version.

**Comment by jancovic** [comment]
hmm, looks all good here

:~/CoinTaxman$ pip install -r requirements.txt 
Requirement already satisfied: certifi==2020.12.5 in /home/.local/lib/python3.8/site-packages (from -r requirements.txt (line 1)) (2020.12.5)
Requirement already satisfied: chardet==4.0.0 in /home/.local/lib/python3.8/site-packages (from -r requirements.txt (line 2)) (4.0.0)
Requirement already satisfied: idna==2.10 in /home/.local/lib/python3.8/site-packages (from -r requirements.txt (line 3)) (2.10)
Requirement already satisfied: python-dateutil==2.8.1 in /home/.local/lib/python3.8/site-packages (from -r requirements.txt (line 4)) (2.8.1)
Requirement already satisfied: requests==2.25.1 in /home/.local/lib/python3.8/site-packages (from -r requirements.txt (line 5)) (2.25.1)
Requirement already satisfied: six==1.15.0 in /home/.local/lib/python3.8/site-packages (from -r requirements.txt (line 6)) (1.15.0)
Requirement already satisfied: urllib3==1.26.5 in /home/.local/lib/python3.8/site-packages (from -r requirements.txt (line 7)) (1.26.5)
$

**Reply by scientes** [reply]
> Requirements
> 
> - Python 3.9

**Comment by jancovic** [comment]
OK, tried again with new 3.9 python , but same error, running on ubuntu 20.04


$ python3.9 --version
Python 3.9.6
 
~/CoinTaxman$ python3.9 "src/main.py"
Traceback (most recent call last):
  File "/home/CoinTaxman/src/main.py", line 20, in <module>
    from book import Book
  File "/home/CoinTaxman/src/book.py", line 25, in <module>
    import config
  File "/home/CoinTaxman/src/config.py", line 20, in <module>
    from dateutil.relativedelta import relativedelta
ModuleNotFoundError: No module named 'dateutil'
~/CoinTaxman$

**Comment by provinzio** [comment]
It's not the same error. It says that you are missing a package. Install the requirements.txt for Python 3.9.

**Comment by jancovic** [comment]
Aaah, you have right, I fix it and finally run it, but got this errors, 
I put csv transaction file to   folder  account_statements/
So it generates no report, right ?

python3.9 src/main.py 
2021-08-31 17:32:23,827 book         INFO     Reading file from exchange binance at /home/CoinTaxman/account_statements/part-00000-1f483d92-acba-4d7e-b8c4-aa4da7083c8e-c000.csv
...
2021-08-31 17:32:23,860 book         WARNING  I may have missed a remark in /home/CoinTaxman/account_statements/part-00000-1f483d92-acba-4d7e-b8c4-aa4da7083c8e-c000.csv:380: `Withdraw fee is included`.
2021-08-31 17:32:23,886 book         WARNING  Could not recognize operation `transfer_out` in  binance file `/home/CoinTaxman/account_statements/part-00000-1f483d92-acba-4d7e-b8c4-aa4da7083c8e-c000.csv:704`.
Traceback (most recent call last):
  File "/home/CoinTaxman/src/main.py", line 44, in <module>
    main()
  File "/home/CoinTaxman/src/main.py", line 32, in main
    status = book.read_files()
  File "/home/CoinTaxman/src/book.py", line 771, in read_files
    self.read_file(file_path)
  File "/home/CoinTaxman/src/book.py", line 726, in read_file
    read_file(file_path)
  File "/home/CoinTaxman/src/book.py", line 112, in _read_binance
    assert account == "Spot", (
AssertionError: Other types than Spot are currently not supported. Please create an Issue or PR.
~/CoinTaxman$

**Reply by scientes** [reply]
haver you traded in something other than spot? margin for instance?

**Reply by scientes** [reply]
look in the csv for rows wich are not labeled spot, future commissions should also count to that. everything not spot is either not implemented or being implemented

**Reply by jancovic** [reply]
ok, but can it ignore other than spot or I need to cut it off ?

**Reply by provinzio** [reply]
It depends. If you traded the same coins on spot and on the unsupported part of binance, you must not cut them off (German Tax law)

**Comment by jancovic** [comment]
yes, I was just  asking to cut off for  skip this error...
Anyway I tried another csv, and this should be really first transactions on this account, and I got another error, I dont know what is problem.

2021-08-31 20:23:12,484 taxman       DEBUG    Starting evaluation...
2021-08-31 20:23:12,486 taxman       ERROR    csv: Line 3: Not enough USDT in queue to sell (transaction from 2021-04-22 08:52:05+00:00 on binance)
        This error occurs if your account statements have unmatched buy/sell positions.
        Have you added all your account statements of the last years?
        This error may also occur after deposits from unknown sources.

Begining of csv>

UTC_Time	Account	Operation	Coin	Change
2021-04-22 01:02:47	Spot	Deposit	USDT	999
2021-04-22 08:52:05	Spot	Transaction Related	USDT	-396.32256712
2021-04-22 08:52:05	Spot	Fee	BTC	-7.39E-06
2021-04-22 08:52:05	Spot	Buy	BTC	0.007391
...

**Reply by provinzio** [reply]
The transaction history is missing the buy of your USDT.

The program can not calculate your gains on the USDT sell on line 3.

**Reply by scientes** [reply]
To calculate taxes, the whole transaction history is needed, deposits and withdrawals are currently ignored, due to not being relevant for the calculation. @provinzio should we mabe not ignore deposits when made in usdt or other fiats?

**Reply by scientes** [reply]
if the usdt deposited here are deposited directly from your bank account (e.g do not come from other exchanges) you could just change that line into a buy operation

**Reply by provinzio** [reply]
> To calculate taxes, the whole transaction history is needed, deposits and withdrawals are currently ignored, due to not being relevant for the calculation. @provinzio should we mabe not ignore deposits when made in usdt or other fiats?

Doesn't hurt to keep it in place. It's ignored for tax calculation but could be used to show the balance on all platforms.

**Comment by jancovic** [comment]
Hi, Sorry for long delay response, I just checked this my old Issue, as end year is coming :)
So on your advice I changed 3rd line operation to BUY, but I got still same error
     1  UTC_Time,Account,Operation,Coin,Change,Remark
     2  2021-04-22 01:02:47,Spot,Deposit,USDT,999.00000000,""
     3  2021-04-22 08:52:05,Spot,Buy,USDT,-396.32256712,""
     4  2021-04-22 08:52:05,Spot,Fee,BTC,-0.00000739,""
     5  2021-04-22 08:52:05,Spot,Buy,BTC,0.00739100,""

2021-12-12 19:49:14,197 taxman       ERROR   transactions_April-Jun_2021.csv: Line 3: Not enough USDT in queue to sell: missing 396.32256712 USDT (transaction from 2021-04-22 08:52:05+00:00 on binance)
        This error occurs if your account statements have unmatched buy/sell positions.
        Have you added all your account statements of the last years?
        This error may also occur after deposits from unknown sources.

**Reply by scientes** [reply]
Line 3 was correct being Transaction related,
change Line 1 to buy (the deposit line)
